### DISCUSSION TITLE
Does WiX toolset automatically do a rollback whenever an error occurs during installation?

### DISCUSSION THREAD

**Discussion by Fiinleyyy** [oleg-shilo/wixsharp#1560] [Q&A] [answered]
Question stated in the title.

**Comment by oleg-shilo** [comment, marked as answer]
yes, though it's not but MSI does the rollback. WiX is a tool for building msi packages. But it is msi runtime that controls the transactional nature of the setup session.

**Comment by Fiinleyyy** [comment]
Thank you. Do you also know what's the best way to execute some lines of code as soon as a Rollback is detected? I could use try/catch but is there a more fancy version?

**Comment by Fiinleyyy** [comment]
edit: followup question

**Comment by oleg-shilo** [comment]
no I don't.
you can use custom you and in the exit dialog, you can detect the setup failure and do something. But it is highly unorthodox for the aborted setup to require any other extra activity. So there is no special mechanism for that.

**Reply by Fiinleyyy** [reply]
The reason I'm asking is because during my installation a powershell script is executed which esentially creates new files/directories. These additional files are of course not uninstalled during the msi uninstall / rollback

**Reply by Fiinleyyy** [reply]
Is there a way I can check (e.g in afterInstall) if the installation was sucessful? If not -> execute function? @oleg-shilo

**Reply by oleg-shilo** [reply]
> Is there a way I can check (e.g in afterInstall) if the installation was sucessful?

Yes, it is possible. You can experiment with this approach:

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/019163fa-b17d-4c47-b95f-ed9763afe36c)

And of course you can also check if the files normally installed by your shell script do exist.

**Comment by Fiinleyyy** [comment]
Thanks for the quick response. I actually had a similiar idea. Upon trying your code unfortunaetely the installation runs into the following error:

> System.IO.FileNotFoundException: Could not load file or assembly 'WixSharp.Msi, Version=2.1.7.0, Culture=neutral, PublicKeyToken=XXXXXXXXXXXXXXXX' or one of its dependencies. The system cannot find the file specified.

This only happens when  following lines are included:

`if (MsiParser.IsInstalled($"{e.ProductCode}"))
{
    MessageBox.Show("Is installed...");
}
else
{
    MessageBox.Show("Rollback...");
}`

My goal eventually is to check if the installation *failed*. If so, execute a function

**Comment by oleg-shilo** [comment]
I see. It means that your setup does not include WixSharp.Msi.dll.

To solve this you need to add it in your setup as a dependency.

```c#
project.DefaultRefAssemblies.Add(typeof(MsiParser).Assembly.Location);
```

**Reply by Fiinleyyy** [reply]
The installer runs now fine! But using your logic the rollback is executed in every of the three scenarios: Install, Uninstall and rollback - Would love for it to only execute on rollback...

**Comment by Fiinleyyy** [comment]
It is also very weird that this is never true even though it's my last code snippet in AfterInstall

`if (MsiParser.IsInstalled($"{e.ProductCode}"))
{
    MessageBox.Show("Is installed...");
}
else
{
    MessageBox.Show("Rollback...");
}`

**Reply by Fiinleyyy** [reply]
I also tried `if(e.Result == ActionResult.Failure)` - but the problem I'm facing is that when I cancel the installation through the wizard ActionResult still stays suceeded

**Comment by Fiinleyyy** [comment]
What is the proper way to implement rollback functionality and is there a way to trigger / test it?

**Comment by oleg-shilo** [comment]
It is probably because AfterInstall is a deferred action and invoked when the install session is over.
I tested it by pressing the cancel button in the progress dialog. But you need to be quick :)

**Reply by Fiinleyyy** [reply]
So with e.Result I can catch if there are any errors in AfterInstall. But how about the installation itself?

**Comment by oleg-shilo** [comment]
You cannot catch `e.Result`. You are the one who is setting it. 
With `e.Result` you can **trigger** rollback you want (e.g `e.Result = ActionResult.UserExit`), but you cannot **catch** rollback.

**Reply by Fiinleyyy** [reply]
Got it. But any other way to "catch" rollback?
